src/timeline/TimelineViewManager.cpp

```diff
@@ -614,3 +618,98 @@ TimelineViewManager::focusTimeline()
 {
         getWidget()->setFocus();
 }
+
+void
+TimelineViewManager::forwardMessageToRoom(mtx::events::collections::TimelineEvents *e,
+                                          QString roomId,
+                                          bool sentFromEncrypted)
+{
+        auto elem        = *e;
+        auto room        = models.find(roomId);
+        auto messageType = mtx::accessors::msg_type(elem);
+
+        if (sentFromEncrypted && messageType == mtx::events::MessageType::Image) {
```

Comment:
This partially applies to all file types. And you don't need to recalculate anything, just upload the file object to generate a an url object instead, everything else can be just taken from the old event. Basically you can just copy the old info object, and replace file with nullopt and set the url instead to the newly uploaded file.

Reply:
Thanks! I've made the changes, should be ready for review now.